autopilot/contracts.go

```diff
@@ -1,16 +1,34 @@
 package autopilot
 
 import (
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
 	"go.sia.tech/renterd/internal/consensus"
+	rhpv2 "go.sia.tech/renterd/rhp/v2"
 	"go.sia.tech/renterd/worker"
 	"go.sia.tech/siad/types"
 	"golang.org/x/crypto/blake2b"
 )
 
+const contractSetName = "autopilot"
+
+func logErr(err error) {} // TODO
+
 func (ap *Autopilot) currentHeight() uint64 {
 	return ap.store.Tip().Height
 }
 
+func (ap *Autopilot) currentPeriod() uint64 {
+	c := ap.store.Config()
+	h := ap.store.Tip().Height
```

Comment:
Tip is a bit of an awkward name for the autopilot store imho. Because it sounds like it would update every 10 minutes.

Also something feels off. Shouldn't the `currentPeriod` be the tip and the `nextPeriod` be `tip + c.Contracts.Period`? Assuming that the latter is the user setting.

Reply:
I think this is leftover from when the autopilot was directly subscribing to consensus. In the new design, the autopilot should get this info from the bus.

Reply:
Hm so I already removed `Tip` from the interface before reading this comment, I don't get this info from the bus however but instead added a `State` variable to the `store` that keeps track of block height, current period and so on. I figured that was ok since the store is subscribed to the consensus.

Reply:
in the new design, only `bus` will subscribe to consensus.

One way to think about autopilot is to imagine it being implemented entirely within the UI. Like, imagine if there was a "enabled autopilot" button in the UI, and when you press it, the mouse and keyboard start moving on their own, clicking on UI buttons and typing stuff in, forming contracts, running host scans, etc. The only information is has access to is the information available in the UI, and the only actions it can perform are the actions available in the UI. In practice, we probably won't be *too* strict about that requirement, but it still serves as a broad guideline for what the autopilot should and shouldn't do. In this case, for example, the autopilot doesn't *directly* know what the current height is; it *learns* the height by "looking at the UI," which in turn is getting the height from `bus`. Hopefully that makes sense 😅

Reply:
I love the visual image you imprinted in my brain about the mouse moving on its own and doing stuff 😛 . We are on the same page though, I just figured it was ok to add `currentPeriod` as state to the autopilot (since I think we agreed on having the config as state in the autopilot) and then decided to just chuck in `synced` and `blockheight` for the time being. Moving it to bus is really just a matter of seconds.

I wanted to explicitly mention the config being stored as state in the autopilot because your UI story is making me think you reconsidered doing that.